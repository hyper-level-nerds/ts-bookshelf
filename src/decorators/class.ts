import { getTypeStorage } from "@utils/type-storage";
import { ClassData } from "@utils/types";

type DocumentTypeOptions = ClassData["userData"];

export function DocType(options?: DocumentTypeOptions): ClassDecorator {
    return target => {
        getTypeStorage().collectClassData({
            classType: target,
            userData: {
                name: options?.name || target.name,
                description: options?.description || `This is a type '${target.name}'.`,
            },
        });
    };
}
