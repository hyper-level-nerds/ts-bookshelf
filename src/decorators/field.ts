import { DecoratorType, FieldData } from "@utils/types";
import { TypeStorage } from "@root/utils/type-storage";

type DocumentFieldOptions = FieldData["userData"];

export function DocField(fieldData: DocumentFieldOptions): DecoratorType {
    return <T>(target: object, propertyKey: string | symbol, descriptor?: TypedPropertyDescriptor<T>) => {
        if (typeof propertyKey === "symbol") {
            throw new Error("Symbol keys are not supported yet!");
        }

        const isMethod = Boolean(descriptor && descriptor.value);
        if (isMethod) {
            throw new Error("Field decorator can only be used on properties!");
        }

        const type = Reflect.getMetadata("design:type", target, propertyKey);
        if (!type || (type !== Number && type !== String && type !== Boolean)) {
            throw new Error("Field decorator can only be used on properties of type Number, String or Boolean!");
        }

        TypeStorage.instance.collectFieldData({
            type,
            fieldName: propertyKey,
            classType: target.constructor,
            userData: {
                ...fieldData,
                description: fieldData.description || `field '${propertyKey}'.`,
                nullable: fieldData.nullable ?? false,
            },
        });
    };
}
