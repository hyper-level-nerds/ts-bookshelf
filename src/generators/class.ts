import { getTypeStorage } from "@utils/type-storage";
import { getHeaderGenerator } from "@utils/generate-header";
import { ClassType } from "@utils/types";

import { generateTableFromField } from "@utils/generate-table";

interface DocumentOptions {
    initialHeaderLevel?: number; // = 1
}

export function generateDocsForClass(classType: ClassType, options?: DocumentOptions) {
    const { initialHeaderLevel = 1 } = options || {};
    const header = getHeaderGenerator(initialHeaderLevel);

    const targetClass = getTypeStorage().classes.find(classData => classData.classType === classType);
    if (!targetClass) {
        throw new Error(`Class '${classType.name}' is not registered!`);
    }

    const contents: string[] = [];
    contents.push(`${header(targetClass.userData.name, 1)} (${targetClass.className})\n`);
    contents.push(`${targetClass.userData.description}\n`);
    contents.push(`${header("Fields", 2)}\n`);

    for (const field of Object.values(targetClass.fieldMap)) {
        contents.push(`${header(`\`${field.fieldName}\``, 3)}\n`);
        contents.push(generateTableFromField(field));
        contents.push("\n\n");
    }

    return contents.join("\n");
}
