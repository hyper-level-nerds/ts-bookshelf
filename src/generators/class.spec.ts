import "reflect-metadata";

import { generateDocsForClass } from "@generators/class";
import { DocField } from "@decorators/field";

describe("generateDocsForClass() Function", function () {
    it("should be defined", function () {
        expect(generateDocsForClass).toBeDefined();
    });

    it("should be a function", function () {
        expect(generateDocsForClass).toBeInstanceOf(Function);
    });

    it("should generate documentation correctly", function () {
        class MockedClass {
            @DocField({
                description: "test",
                nullable: false,
            })
            public test!: boolean;

            @DocField({
                description: "test",
                nullable: true,
            })
            public test2!: boolean;
        }

        const generatedData = generateDocsForClass(MockedClass);

        expect(generatedData).toMatchSnapshot();
    });

    it("should throw an error if the class is not registered", function () {
        expect(() => {
            generateDocsForClass(String);
        }).toThrowError("Class 'String' is not registered!");
    });
});
