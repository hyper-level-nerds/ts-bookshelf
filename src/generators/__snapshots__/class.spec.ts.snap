// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDocsForClass() Function should generate documentation correctly 1`] = `
"# MockedClass (MockedClass)

type 'MockedClass'.

## Fields

### \`test\`

| Name        | Description |
| ----------- | ----------- |
| Type        | Boolean     |
| Nullable    | ❌ No        |
| Description | test        |



### \`test2\`

| Name        | Description |
| ----------- | ----------- |
| Type        | Boolean     |
| Nullable    | ✔️ Yes      |
| Description | test        |


"
`;
